sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -856,19 +867,65 @@ def waiting_time_compensation_with_sleep(time_start, time_to_wait):
     if time_diff < time_to_wait:
         time.sleep(time_to_wait - time_diff)
 
-# Update port SFP status table on receiving SFP change event
+# Update port SFP status table for SW fields on receiving SFP change event
 
 
-def update_port_transceiver_status_table(logical_port_name, status_tbl, status, error_descriptions='N/A'):
+def update_port_transceiver_status_table_sw(logical_port_name, status_tbl, status, error_descriptions='N/A'):
     fvs = swsscommon.FieldValuePairs([('status', status), ('error', error_descriptions)])
     status_tbl.set(logical_port_name, fvs)
 
+# Update port SFP status table for HW fields
+
+
+def update_port_transceiver_status_table_hw(logical_port_name, port_mapping,
+                                            table, stop_event=threading.Event(), transceiver_status_cache=None):
+    for physical_port, physical_port_name in get_physical_port_name_dict(logical_port_name, port_mapping).items():
+        if stop_event.is_set():
+            break
+
+        if not _wrapper_get_presence(physical_port):
+            continue
+
+        try:
+            if transceiver_status_cache is not None and physical_port in transceiver_status_cache:
+                # If cache is enabled and status info is in cache, just read from cache, no need read from EEPROM
+                transceiver_status_dict = transceiver_status_cache[physical_port]
+            else:
+                transceiver_status_dict = _wrapper_get_transceiver_status(physical_port)
+                if transceiver_status_cache is not None:
+                    # If cache is enabled, put status info to cache
+                    transceiver_status_cache[physical_port] = transceiver_status_dict
+            if transceiver_status_dict is not None:
+                beautify_transceiver_status_dict(transceiver_status_dict, physical_port)
+                fvs = swsscommon.FieldValuePairs([(k, v) for k, v in transceiver_status_dict.items()])
+                table.set(physical_port_name, fvs)
+            else:
+                return SFP_EEPROM_NOT_READY
+
+        except NotImplementedError:
+            helper_logger.log_error("get_transceiver_status is currently not implemented for this platform")
+            sys.exit(NOT_IMPLEMENTED_ERROR)
 
 # Delete port from SFP status table
 
 
-def delete_port_from_status_table(logical_port_name, status_tbl):
-    status_tbl._del(logical_port_name)
+def delete_port_from_status_table_sw(logical_port_name, status_tbl):
+    for f in TRANSCEIVER_STATUS_TABLE_SW_FIELDS:
+        status_tbl.hdel(logical_port_name, f)
+
+# Delete port from SFP status table for HW fields which are fetched from EEPROM
+
+
+def delete_port_from_status_table_hw(logical_port_name, port_mapping, status_tbl):
+    for physical_port_name in get_physical_port_name_dict(logical_port_name, port_mapping).values():
+        found, fvs = status_tbl.get(physical_port_name)
```

Comment:
why use **physical_port_name** as the index and not **logical_port_name**?

Reply:
Because this function is supposed to only delete the HW fields of status tbl, which are fetched from eeprom (more precisely VDM) and per physical port.

Reply:
I believe most of the current xcvrd relevant fileds in the state DB are fetching from the hw(EEPROM), but all of them are using  'logical_port_name' as the index, there will be duplicated information in the DB(multiple logical_port can be created from the same physical port), but the user can easily query them from the CLI/SNMP with the logical port name. I suggest to follow the same convention here.  @prgeor please also share your view here.

Reply:
AFAIK, SW part (fields 'status' and 'error') of status table are not from EEPROM, and all the other fields on status table (introduced by this PR) and other tables are from EEPROM. 
And currently, only SW part of status table is using logical_port_name as key, all the other existing tables (dom_sensor tbl/info tbl, those are eeprom-feteched) are using physical_port_name as index. I'm just making the HW part of status table aligned with other existing tables.

e.g. in today’s post_port_sfp_info_to_db(), post_port_dom_threshold_info_to_db() and post_port_dom_info_to_db(), they all have the following code to use physical port to set DB:
    for physical_port in physical_port_list:
        …..
        port_name = get_physical_port_name(logical_port_name, ganged_member_num, ganged_port)
        …..
        table.set(port_name, fvs)

We discussed with @prgeor about this before, and we agreed to better use logical_port_name as key, and https://github.com/sonic-net/sonic-buildimage/issues/10144 can be used to track this pending work item. But the plan was to keep this item as lower priority, since it may only cause issue on DPB case for now. And also the history/necessarity of ganged port needs to figured out first (ganged port can have 1:n mapping from logical port to physical port), so that we can decide wether the ganged port related logic can be cleaned up around this area.

Reply:
> AFAIK, SW part (fields 'status' and 'error') of status table are not from EEPROM, and all the other fields on status table (introduced by this PR) and other tables are from EEPROM. And currently, only SW part of status table is using logical_port_name as key, all the other existing tables (dom_sensor tbl/info tbl, those are eeprom-feteched) are using physical_port_name as index. I'm just making the HW part of status table aligned with other existing tables.

> 
> e.g. in today’s post_port_sfp_info_to_db(), post_port_dom_threshold_info_to_db() and post_port_dom_info_to_db(), they all have the following code to use physical port to set DB:
> 
> ```
>     for physical_port in physical_port_list:
>         …..
>         port_name = get_physical_port_name(logical_port_name, ganged_member_num, ganged_port)
>         …..
>         table.set(port_name, fvs)
> ```
>

This is not true, check below code you will see that for all none-ganged ports it will always use logical port name including table DOM and Transceiver info table. By the way, I haven't seen any vendor claiming that they have ganged port on any of their platforms.

```
def get_physical_port_name(logical_port, physical_port, ganged):
    if ganged:
        return logical_port + ":{} (ganged)".format(physical_port)
    else:
        return logical_port
```
 
> We discussed with @prgeor about this before, and we agreed to better use logical_port_name as key, and [sonic-net/sonic-buildimage#10144](https://github.com/sonic-net/sonic-buildimage/issues/10144) can be used to track this pending work item to clean up all the existing code to use logical_port_name as key for DB tables. But the plan was to keep this item as lower priority, since it may only cause issue on DPB case for now. And also the history/necessarity of ganged port needs to be figured out first (ganged port can have 1:n mapping from logical port to physical port), so that we can decide wether the ganged port related logic can be cleaned up around this area. Also, to clean this area up, more consideration/discussion would be needed later on whether we need to divide the corresponding eeprom-fetched fields belonging to the same physical port into multiple groups based on logical ports. e.g. for the following fields in TRANSCEIVER_INFO table, if we are in breakout mode, each logical port will only have part of the 8 lanes, we might need to divide them based on logical port.
>

If you change the index of the tables the CLIs (e.g 'show interface transceiver eeprom') will be broken, and also SNMP and other telemetry applications assuming the index shall be logical port index will also be broken, please double confirm.
 
> ```
>         active_apsel_hostlane1       = INTEGER                  ; active application selected code assigned to host lane 1
>         active_apsel_hostlane2       = INTEGER                  ; active application selected code assigned to host lane 2
>         active_apsel_hostlane3       = INTEGER                  ; active application selected code assigned to host lane 3
>         active_apsel_hostlane4       = INTEGER                  ; active application selected code assigned to host lane 4
>         active_apsel_hostlane5       = INTEGER                  ; active application selected code assigned to host lane 5
>         active_apsel_hostlane6       = INTEGER                  ; active application selected code assigned to host lane 6
>         active_apsel_hostlane7       = INTEGER                  ; active application selected code assigned to host lane 7
>         active_apsel_hostlane8       = INTEGER                  ; active application selected code assigned to host lane 8
> ```